src/util/ALE.sol

```diff
@@ -0,0 +1,459 @@
+    error DOLAInvalidBorrow(uint256 expected, uint256 actual);
+    error DOLAInvalidRepay(uint256 expected, uint256 actual);
+    error InvalidProxyAddress();
+    error InvalidHelperAddress();
+    error NothingToDeposit();
+    error DepositFailed(uint256 expected, uint256 actual);
+    error WithdrawFailed(uint256 expected, uint256 actual);
+    error TotalSupplyChanged(uint256 expected, uint256 actual);
+
+    // 0x ExchangeProxy address.
+    // See https://docs.0x.org/developer-resources/contract-addresses
+    address payable public exchangeProxy;
+
+    struct Market {
+        IERC20 buySellToken;
+        IERC20 collateral;
+        ITransformHelper helper;
+    }
+
+    struct Permit {
+        uint256 deadline;
+        uint8 v;
+        bytes32 r;
+        bytes32 s;
+    }
+
+    struct DBRHelper {
+        uint256 amountIn; // DOLA or DBR
+        uint256 minOut; // DOLA or DBR
+        uint256 dola; // DOLA to extra borrow or extra repay
+    }
+
+    event LeverageUp(
+        address indexed market,
+        address indexed account,
+        uint256 dolaFlashMinted,
+        uint256 collateralBought
+    );
+
+    event LeverageDown(
+        address indexed market,
+        address indexed account,
+        uint256 dolaFlashMinted,
+        uint256 collateralSold
+    );
+
+    event NewMarket(address indexed market, address indexed buySellToken, address collateral, address indexed helper);
+
+    event NewHelper(address indexed market, address indexed helper);
+
+    // Mapping of market to Market structs
+    // NOTE: in normal cases sellToken/buyToken is the collateral token,
+    // in other cases it could be different (eg. st-yCRV is collateral, yCRV is the token to be swapped from/to DOLA)
+    mapping(address => Market) public markets;
+
+    modifier dolaSupplyUnchanged() {
+        uint256 totalSupply = dola.totalSupply();
+        _;
+        if (totalSupply != dola.totalSupply())
+            revert TotalSupplyChanged(totalSupply, dola.totalSupply());
+    }
+
+    constructor(
+        address _exchangeProxy,
+        address _pool
+    ) Ownable(msg.sender) CurveDBRHelper(_pool) {
+        exchangeProxy = payable(address(_exchangeProxy));
+    }
+
+    function setExchangeProxy(address _exchangeProxy) external onlyOwner {
+        if (_exchangeProxy == address(0)) revert InvalidProxyAddress();
+        exchangeProxy = payable(_exchangeProxy);
+    }
+
+    /// @notice Set the market for a collateral token
+    /// @param _buySellToken The token which will be bought/sold (usually the collateral token), probably underlying if there's a helper
+    /// @param _market The market contract
+    /// @param _collateral The collateral token
+    /// @param _helper Optional helper contract to transform collateral to buySelltoken and viceversa
+    function setMarket(
+        address _market,
+        address _buySellToken,
+        address _collateral,
+        address _helper
+    ) external onlyOwner {
```

Comment:
Suggestion:
Add input validation against `DBR.markets()` to verify that this is a valid market.
Similarly , can ensure that the `_collateral` input matches `market.collateral()`